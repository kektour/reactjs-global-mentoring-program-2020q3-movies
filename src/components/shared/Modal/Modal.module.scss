@import 'src/shared.scss';

.rootOpen {
  filter: blur(10px);
}

.bodyOpen {
  overflow: hidden;
}

.root {
  top: 0;
  bottom: 0;
  left: 0;
  right: 0;
  position: fixed;
  z-index: 9999;
  overflow-y: auto;
  height: 100%;
  display: flex;
  flex-direction: column;
  background: rgba(0, 0, 0, 0.4);
  box-shadow: inset 120px 100px 250px #000000, inset -120px -100px 250px #000000;
}

.container {
  @extend %container;
  flex: 1 0 auto;
  margin-bottom: 125px;
}

.header {
  padding-top: 20px;
  margin-bottom: 80px;
}

.footerRoot {
  flex-shrink: 0;
}
