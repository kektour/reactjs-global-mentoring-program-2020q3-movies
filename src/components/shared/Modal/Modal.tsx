import React, { useMemo } from 'react';
import ReactDOM from 'react-dom';
import { Footer } from '../Footer';
import { Header } from '../Header';
import styles from './Modal.module.scss';

type Props = {
  open: boolean;
  children: React.ReactNode;
};

export const Modal: React.FC<Props> = (props) => {
  const { open, children } = props;
  const body = document.querySelector('body') as HTMLBodyElement;
  const root = document.querySelector('#root') as HTMLDivElement;
  useMemo(() => {
    if (open) {
      body.classList.add(styles.bodyOpen);
      root.classList.add(styles.rootOpen);
    } else {
      body.classList.remove(styles.bodyOpen);
      root.classList.remove(styles.rootOpen);
    }
  }, [open, body, root]);

  return (
    (open && (
      <React.Fragment>
        {ReactDOM.createPortal(
          <div className={styles.root}>
            <Header classes={{ root: styles.header }} />
            <div className={styles.container}>
              {children}
            </div>
            <Footer classes={{ root: styles.footerRoot }} />
          </div>,
          body
        )}
      </React.Fragment>
    )) ||
    null
  );
};
