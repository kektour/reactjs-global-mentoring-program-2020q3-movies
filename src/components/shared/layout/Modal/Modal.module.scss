.root {
  top: 0;
  bottom: 0;
  left: 0;
  right: 0;
  position: fixed;
  z-index: 9999;
  overflow-y: auto;
}

.bodyOpen {
  overflow: hidden;
}
