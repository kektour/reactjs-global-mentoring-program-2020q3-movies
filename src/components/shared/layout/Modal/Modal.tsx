import React, { useMemo } from 'react';
import ReactDOM from 'react-dom';
import classnames from 'classnames';
import styles from './Modal.module.scss';

type Props = {
  open: boolean;
  children: React.ReactNode;
  classes?: {
    root?: string;
  };
};

export const Modal: React.FC<Props> = (props) => {
  const { open, children, classes = {} } = props;
  const body = useMemo(() => document.querySelector('body') as HTMLBodyElement, []);
  useMemo(() => {
    if (open) {
      body.classList.add(styles.bodyOpen);
    } else {
      body.classList.remove(styles.bodyOpen);
    }
  }, [open]);

  return (
    (open && (
      <React.Fragment>
        {ReactDOM.createPortal(<div className={classnames(styles.root, classes.root)}>{children}</div>, body)}
      </React.Fragment>
    )) ||
    null
  );
};
