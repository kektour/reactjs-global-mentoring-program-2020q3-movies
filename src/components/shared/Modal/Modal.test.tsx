import React from 'react';
import ReactDOM from 'react-dom';

import { render } from '@testing-library/react';
import { Modal } from './Modal';

describe('Modal', () => {
  const spyQuerySelector = jest.fn();

  beforeAll(() => {
    Object.defineProperty(global.document, 'querySelector', { value: spyQuerySelector });
    ReactDOM.createPortal = jest.fn((e) => e) as any;
  });

  it('Should render component correctly', () => {
    const body = document.createElement('body');
    const div = document.createElement('div');
    spyQuerySelector.mockImplementationOnce(() => body);
    spyQuerySelector.mockImplementationOnce(() => div);

    const component = render(
      <Modal open={true}>
        <div>Foo</div>
      </Modal>
    );
    expect(component.asFragment()).toMatchSnapshot();
  });
});
