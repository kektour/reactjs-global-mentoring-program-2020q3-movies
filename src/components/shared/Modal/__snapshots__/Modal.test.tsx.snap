// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal Should render component correctly 1`] = `
<DocumentFragment>
  <div
    class="root"
  >
    <div
      class="root header"
    >
      <div
        class="container"
      >
        <div
          class="header"
        >
          <p
            class="root"
          >
            <span>
              netflix
            </span>
            roulette
          </p>
        </div>
      </div>
    </div>
    <div
      class="container"
    >
      <div>
        Foo
      </div>
    </div>
    <div
      class="root footerRoot"
    >
      <p
        class="root logoRoot"
      >
        <span>
          netflix
        </span>
        roulette
      </p>
    </div>
  </div>
</DocumentFragment>
`;
